In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import plotly.express as px


%matplotlib inline

df_final_demo = pd.read_csv("../01_data/df_final_demo.txt")
df_final_demo

df_final_experiment_clients = pd.read_csv("../01_data/df_final_experiment_clients.txt")
df_final_experiment_clients

df_final_web_data_pt_1 = pd.read_csv("../01_data/df_final_web_data_pt_1.txt")
df_final_web_data_pt_1

df_final_web_data_pt_2 = pd.read_csv("../01_data/df_final_web_data_pt_2.txt")
df_final_web_data_pt_2

df_final_web_data = pd.concat([df_final_web_data_pt_1,df_final_web_data_pt_2], axis= 0)
df_final_web_data

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
412259    9668240   388766751_9038881013   922267647_3096648104_968866   
412260    9668240   388766751_9038881013   922267647_3096648104_968866   
412261    9668240   388766751_9038881013   922267647_3096648104_968866   
412262    9668240   388766751_9038881013   922267647_3096648104_968866   
412263     674799  947159805_81558194550   86152093_47511127657_716022   

       process_step            date_time  
0            step_3  2017-04-17 15:27:07  
1            step_2  2017-04-17 15:26:51  
2            step_3  2017-04-17 15:19:22  
3            step_2  2017-04-17 15:19:13  
4            step_3  2017-04-17 15:18:04  
...             ...                  ...  
412259        start  2017-05-24 18:46:10  
412260        start  2017-05-24 18:45:29  
412261       step_1  2017-05-24 18:44:51  
412262        start  2017-05-24 18:44:34  
412263        start  2017-06-03 12:17:09  

[755405 rows x 5 columns]

In [2]:
# This is the data set that contains the demograohic data of the clients.
df_final_demo.head(10)

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0     836976             6.0              73.0      60.5     U        2.0   
1    2304905             7.0              94.0      58.0     U        2.0   
2    1439522             5.0              64.0      32.0     U        2.0   
3    1562045            16.0             198.0      49.0     M        2.0   
4    5126305            12.0             145.0      33.0     F        2.0   
5    3727881             5.0              71.0      30.5     U        2.0   
6     272934             5.0              66.0      58.5     U        2.0   
7     388801            30.0             361.0      57.5     M        5.0   
8     285619            30.0             369.0      67.5     M        2.0   
9    8198645            15.0             189.0      54.5     F        2.0   

         bal  calls_6_mnth  logons_6_mnth  
0   45105.30           6.0            9.0  
1  110860.30           6.0            9.0  
2   52467.79           6.0            9.0  
3   67454.65           3.0            6.0  
4  103671.75           0.0            3.0  
5   23915.60           0.0            3.0  
6   27021.42           2.0            5.0  
7  522498.72           1.0            4.0  
8  299388.72           3.0            6.0  
9  382303.83           6.0            9.0

In [3]:
df_final_demo.isnull().sum()

client_id            0
clnt_tenure_yr      14
clnt_tenure_mnth    14
clnt_age            15
gendr               14
num_accts           14
bal                 14
calls_6_mnth        14
logons_6_mnth       14
dtype: int64

In [4]:
# checking the two columns for nulls so that we can drop them
df_final_demo[df_final_demo['clnt_age'].isnull() | df_final_demo['gendr'].isnull()] 

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
4164     7402828             NaN               NaN       NaN   NaN        NaN   
8316      355337             NaN               NaN       NaN   NaN        NaN   
8677     8412164             NaN               NaN       NaN   NaN        NaN   
9583     4666211             8.0             106.0       NaN     F        2.0   
13444    2222915             NaN               NaN       NaN   NaN        NaN   
18066    4876926             NaN               NaN       NaN   NaN        NaN   
25961    5277910             NaN               NaN       NaN   NaN        NaN   
28432    7616759             NaN               NaN       NaN   NaN        NaN   
35323    8191345             NaN               NaN       NaN   NaN        NaN   
43518    1227228             NaN               NaN       NaN   NaN        NaN   
46076    8611797             NaN               NaN       NaN   NaN        NaN   
47189    5144725             NaN               NaN       NaN   NaN        NaN   
49846    1037867             NaN               NaN       NaN   NaN        NaN   
63701    1618749             NaN               NaN       NaN   NaN        NaN   
66219    6021001             NaN               NaN       NaN   NaN        NaN   

            bal  calls_6_mnth  logons_6_mnth  
4164        NaN           NaN            NaN  
8316        NaN           NaN            NaN  
8677        NaN           NaN            NaN  
9583   42550.55           4.0            7.0  
13444       NaN           NaN            NaN  
18066       NaN           NaN            NaN  
25961       NaN           NaN            NaN  
28432       NaN           NaN            NaN  
35323       NaN           NaN            NaN  
43518       NaN           NaN            NaN  
46076       NaN           NaN            NaN  
47189       NaN           NaN            NaN  
49846       NaN           NaN            NaN  
63701       NaN           NaN            NaN  
66219       NaN           NaN            NaN

In [5]:
# drop the rows with all nulls
df_final_demo = df_final_demo.dropna( subset = ['clnt_tenure_yr'])

In [6]:
df_final_demo.isnull().sum()

client_id           0
clnt_tenure_yr      0
clnt_tenure_mnth    0
clnt_age            1
gendr               0
num_accts           0
bal                 0
calls_6_mnth        0
logons_6_mnth       0
dtype: int64

In [7]:
df_final_experiment_clients. head(10)

client_id Variation
0    9988021      Test
1    8320017      Test
2    4033851   Control
3    1982004      Test
4    9294070   Control
5    9466803      Test
6    3170143      Test
7    6651403   Control
8    5887227      Test
9    2105948   Control

In [8]:
df_final_web_data_pt_1.head(10)

client_id            visitor_id                      visit_id process_step  \
0    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
1    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
2    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
3    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
4    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
5    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
6    9988021  580560515_7732621733  781255054_21935453173_531117       step_1   
7    9988021  580560515_7732621733  781255054_21935453173_531117        start   
8    8320017  39393514_33118319366  960651974_70596002104_312201      confirm   
9    8320017  39393514_33118319366  960651974_70596002104_312201       step_3   

             date_time  
0  2017-04-17 15:27:07  
1  2017-04-17 15:26:51  
2  2017-04-17 15:19:22  
3  2017-04-17 15:19:13  
4  2017-04-17 15:18:04  
5  2017-04-17 15:17:15  
6  2017-04-17 15:17:01  
7  2017-04-17 15:16:22  
8  2017-04-05 13:10:05  
9  2017-04-05 13:09:43

In [9]:
df_final_web_data_pt_2.head(10)

client_id             visitor_id                      visit_id  \
0     763412  601952081_10457207388  397475557_40440946728_419634   
1    6019349  442094451_91531546617  154620534_35331068705_522317   
2    6019349  442094451_91531546617  154620534_35331068705_522317   
3    6019349  442094451_91531546617  154620534_35331068705_522317   
4    6019349  442094451_91531546617  154620534_35331068705_522317   
5    6019349  442094451_91531546617  154620534_35331068705_522317   
6    4726500  934350987_45569789638  467318052_88159801968_565608   
7    4726500  934350987_45569789638  467318052_88159801968_565608   
8    4726500  934350987_45569789638  467318052_88159801968_565608   
9    4726500  934350987_45569789638  467318052_88159801968_565608   

  process_step            date_time  
0      confirm  2017-06-06 08:56:00  
1      confirm  2017-06-01 11:59:27  
2       step_3  2017-06-01 11:58:48  
3       step_2  2017-06-01 11:58:08  
4       step_1  2017-06-01 11:57:58  
5        start  2017-06-01 11:57:54  
6      confirm  2017-06-05 17:38:52  
7       step_3  2017-06-05 17:38:33  
8       step_2  2017-06-05 17:37:31  
9       step_1  2017-06-05 17:37:24

In [10]:
df_final_experiment_clients.isnull().value_counts()

#drop null values as this information is decisive for our KPIs
df_final_experiment_clients = df_final_experiment_clients.dropna(subset=['Variation'])
df_final_experiment_clients

client_id Variation
0        9988021      Test
1        8320017      Test
2        4033851   Control
3        1982004      Test
4        9294070   Control
...          ...       ...
50495     393005   Control
50496    2908510   Control
50497    7230446      Test
50498    5230357      Test
50499    6334360      Test

[50500 rows x 2 columns]

In [11]:
#Merge the variation and experiment data sets

df_final_web_data_client_id = pd.merge(df_final_experiment_clients,df_final_web_data, on = 'client_id', how = 'inner')
df_final_web_data_client_id

client_id Variation             visitor_id  \
0         9988021      Test   580560515_7732621733   
1         9988021      Test   580560515_7732621733   
2         9988021      Test   580560515_7732621733   
3         9988021      Test   580560515_7732621733   
4         9988021      Test   580560515_7732621733   
...           ...       ...                    ...   
321304    6334360      Test  629124187_65258232847   
321305    6334360      Test  629124187_65258232847   
321306    6334360      Test  629124187_65258232847   
321307    6334360      Test  629124187_65258232847   
321308    6334360      Test  629124187_65258232847   

                            visit_id process_step            date_time  
0       781255054_21935453173_531117       step_3  2017-04-17 15:27:07  
1       781255054_21935453173_531117       step_2  2017-04-17 15:26:51  
2       781255054_21935453173_531117       step_3  2017-04-17 15:19:22  
3       781255054_21935453173_531117       step_2  2017-04-17 15:19:13  
4       781255054_21935453173_531117       step_3  2017-04-17 15:18:04  
...                              ...          ...                  ...  
321304  586047816_14599436811_389185       step_2  2017-04-11 08:45:38  
321305  586047816_14599436811_389185       step_1  2017-04-11 08:44:43  
321306  586047816_14599436811_389185       step_2  2017-04-11 08:44:30  
321307  586047816_14599436811_389185       step_1  2017-04-11 08:42:36  
321308  586047816_14599436811_389185        start  2017-04-11 08:42:20  

[321309 rows x 6 columns]

In [12]:
# final test group - for KPI

df_final_test = df_final_web_data_client_id[df_final_web_data_client_id['Variation'] == 'Test']
df_final_test

client_id Variation             visitor_id  \
0         9988021      Test   580560515_7732621733   
1         9988021      Test   580560515_7732621733   
2         9988021      Test   580560515_7732621733   
3         9988021      Test   580560515_7732621733   
4         9988021      Test   580560515_7732621733   
...           ...       ...                    ...   
321304    6334360      Test  629124187_65258232847   
321305    6334360      Test  629124187_65258232847   
321306    6334360      Test  629124187_65258232847   
321307    6334360      Test  629124187_65258232847   
321308    6334360      Test  629124187_65258232847   

                            visit_id process_step            date_time  
0       781255054_21935453173_531117       step_3  2017-04-17 15:27:07  
1       781255054_21935453173_531117       step_2  2017-04-17 15:26:51  
2       781255054_21935453173_531117       step_3  2017-04-17 15:19:22  
3       781255054_21935453173_531117       step_2  2017-04-17 15:19:13  
4       781255054_21935453173_531117       step_3  2017-04-17 15:18:04  
...                              ...          ...                  ...  
321304  586047816_14599436811_389185       step_2  2017-04-11 08:45:38  
321305  586047816_14599436811_389185       step_1  2017-04-11 08:44:43  
321306  586047816_14599436811_389185       step_2  2017-04-11 08:44:30  
321307  586047816_14599436811_389185       step_1  2017-04-11 08:42:36  
321308  586047816_14599436811_389185        start  2017-04-11 08:42:20  

[177847 rows x 6 columns]

In [13]:
# final control group - for the KPI's

df_final_control = df_final_web_data_client_id[df_final_web_data_client_id['Variation'] == 'Control']
df_final_control

client_id Variation             visitor_id  \
20        4033851   Control  762728880_76361333336   
21        4033851   Control  762728880_76361333336   
22        4033851   Control  762728880_76361333336   
23        4033851   Control  762728880_76361333336   
24        4033851   Control  762728880_76361333336   
...           ...       ...                    ...   
321265     393005   Control  238199953_33257685972   
321266    2908510   Control  814969699_90652851448   
321267    2908510   Control  814969699_90652851448   
321268    2908510   Control  814969699_90652851448   
321269    2908510   Control  814969699_90652851448   

                            visit_id process_step            date_time  
20      949661017_22392791362_127391      confirm  2017-04-05 12:29:03  
21      949661017_22392791362_127391      confirm  2017-04-05 12:29:01  
22      949661017_22392791362_127391      confirm  2017-04-05 12:28:52  
23      949661017_22392791362_127391       step_3  2017-04-05 12:26:08  
24      949661017_22392791362_127391       step_2  2017-04-05 12:24:43  
...                              ...          ...                  ...  
321265  579367767_60788307935_908431        start  2017-03-29 15:25:20  
321266   562606085_36368381773_92090        start  2017-05-10 22:57:17  
321267   562606085_36368381773_92090       step_2  2017-05-10 22:56:31  
321268   562606085_36368381773_92090       step_1  2017-05-10 22:56:23  
321269   562606085_36368381773_92090        start  2017-05-10 22:56:20  

[143462 rows x 6 columns]

In [14]:
df_final_demo.shape



(70595, 9)

In [15]:
#merge final demo that has the variations with final experiment clients datasets on client_id
# MASTER DEMOGRAPHIC DATASET merged with the ids

df_final_demo_client_id = pd.merge(df_final_experiment_clients,df_final_demo, on = 'client_id', how = 'inner')

df_final_demo_client_id

client_id Variation  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  \
0        9988021      Test             5.0              64.0      79.0     U   
1        8320017      Test            22.0             274.0      34.5     M   
2        4033851   Control            12.0             149.0      63.5     M   
3        1982004      Test             6.0              80.0      44.5     U   
4        9294070   Control             5.0              70.0      29.0     U   
...          ...       ...             ...               ...       ...   ...   
50483     393005   Control            15.0             191.0      52.5     M   
50484    2908510   Control            21.0             252.0      34.0     M   
50485    7230446      Test             6.0              74.0      62.0     M   
50486    5230357      Test            23.0             278.0      30.5     M   
50487    6334360      Test             4.0              53.0      66.5     U   

       num_accts        bal  calls_6_mnth  logons_6_mnth  
0            2.0  189023.86           1.0            4.0  
1            2.0   36001.90           5.0            8.0  
2            2.0  142642.26           5.0            8.0  
3            2.0   30231.76           1.0            4.0  
4            2.0   34254.54           0.0            3.0  
...          ...        ...           ...            ...  
50483        2.0   60344.67           1.0            4.0  
50484        3.0  141808.05           6.0            9.0  
50485        2.0   58778.11           2.0            5.0  
50486        2.0   61349.70           0.0            3.0  
50487        2.0   93103.47           6.0            9.0  

[50488 rows x 10 columns]

In [16]:
df_final_demo_client_id['gendr']=df_final_demo_client_id['gendr'].replace('X','U')
df_final_demo_client_id['gendr'].value_counts(normalize=True)

gendr
U    0.342299
M    0.335664
F    0.322037
Name: proportion, dtype: float64

In [17]:
df_final_demo_client_id.describe()

client_id  clnt_tenure_yr  clnt_tenure_mnth      clnt_age  \
count  5.048800e+04    50488.000000      50488.000000  50487.000000   
mean   5.006173e+06       12.031730        150.415485     47.319795   
std    2.877417e+06        6.860282         81.944830     15.518616   
min    5.550000e+02        2.000000         33.000000     17.000000   
25%    2.515700e+06        6.000000         82.000000     33.500000   
50%    5.025026e+06       11.000000        136.000000     48.000000   
75%    7.477918e+06       16.000000        192.000000     59.500000   
max    9.999832e+06       55.000000        669.000000     96.000000   

          num_accts           bal  calls_6_mnth  logons_6_mnth  
count  50488.000000  5.048800e+04  50488.000000   50488.000000  
mean       2.254575  1.495147e+05      3.093289       6.131873  
std        0.533671  3.020364e+05      2.187991       2.175423  
min        1.000000  2.378944e+04      0.000000       3.000000  
25%        2.000000  3.987841e+04      1.000000       4.000000  
50%        2.000000  6.573360e+04      3.000000       6.000000  
75%        2.000000  1.399565e+05      5.000000       8.000000  
max        7.000000  1.632004e+07      6.000000       9.000000

In [18]:


df_final_demo_control = df_final_demo_client_id[df_final_demo_client_id['Variation'] == 'Control']

df_final_demo_control.describe()


client_id  clnt_tenure_yr  clnt_tenure_mnth      clnt_age  \
count  2.352700e+04    23527.000000      23527.000000  23526.000000   
mean   5.013661e+06       12.087686        151.060186     47.498427   
std    2.877353e+06        6.877531         82.126443     15.521781   
min    1.028000e+03        2.000000         33.000000     17.000000   
25%    2.515289e+06        6.000000         82.000000     33.500000   
50%    5.023798e+06       11.000000        137.000000     48.500000   
75%    7.483504e+06       16.000000        192.000000     60.000000   
max    9.998921e+06       55.000000        669.000000     96.000000   

          num_accts           bal  calls_6_mnth  logons_6_mnth  
count  23527.000000  2.352700e+04  23527.000000   23527.000000  
mean       2.259914  1.501473e+05      3.129213       6.166277  
std        0.538559  2.958070e+05      2.184502       2.173014  
min        2.000000  2.378961e+04      0.000000       3.000000  
25%        2.000000  4.012335e+04      1.000000       4.000000  
50%        2.000000  6.602418e+04      3.000000       6.000000  
75%        2.000000  1.406060e+05      5.000000       8.000000  
max        7.000000  8.292996e+06      6.000000       9.000000

In [19]:

df_final_demo_test = df_final_demo_client_id[df_final_demo_client_id['Variation'] == 'Test']

df_final_demo_test.describe()

client_id  clnt_tenure_yr  clnt_tenure_mnth      clnt_age  \
count  2.696100e+04    26961.000000      26961.000000  26961.000000   
mean   4.999639e+06       11.982901        149.852899     47.163922   
std    2.877510e+06        6.844947         81.783390     15.514461   
min    5.550000e+02        2.000000         33.000000     17.000000   
25%    2.516570e+06        6.000000         82.000000     33.500000   
50%    5.025928e+06       11.000000        134.000000     47.500000   
75%    7.470213e+06       15.000000        191.000000     59.500000   
max    9.999832e+06       55.000000        669.000000     96.000000   

          num_accts           bal  calls_6_mnth  logons_6_mnth  
count  26961.000000  2.696100e+04  26961.000000   26961.000000  
mean       2.249917  1.489626e+05      3.061941       6.101851  
std        0.529335  3.073737e+05      2.190590       2.177119  
min        1.000000  2.378944e+04      0.000000       3.000000  
25%        2.000000  3.972920e+04      1.000000       4.000000  
50%        2.000000  6.546836e+04      3.000000       6.000000  
75%        2.000000  1.394786e+05      5.000000       8.000000  
max        7.000000  1.632004e+07      6.000000       9.000000

In [20]:
df_final_test

client_id Variation             visitor_id  \
0         9988021      Test   580560515_7732621733   
1         9988021      Test   580560515_7732621733   
2         9988021      Test   580560515_7732621733   
3         9988021      Test   580560515_7732621733   
4         9988021      Test   580560515_7732621733   
...           ...       ...                    ...   
321304    6334360      Test  629124187_65258232847   
321305    6334360      Test  629124187_65258232847   
321306    6334360      Test  629124187_65258232847   
321307    6334360      Test  629124187_65258232847   
321308    6334360      Test  629124187_65258232847   

                            visit_id process_step            date_time  
0       781255054_21935453173_531117       step_3  2017-04-17 15:27:07  
1       781255054_21935453173_531117       step_2  2017-04-17 15:26:51  
2       781255054_21935453173_531117       step_3  2017-04-17 15:19:22  
3       781255054_21935453173_531117       step_2  2017-04-17 15:19:13  
4       781255054_21935453173_531117       step_3  2017-04-17 15:18:04  
...                              ...          ...                  ...  
321304  586047816_14599436811_389185       step_2  2017-04-11 08:45:38  
321305  586047816_14599436811_389185       step_1  2017-04-11 08:44:43  
321306  586047816_14599436811_389185       step_2  2017-04-11 08:44:30  
321307  586047816_14599436811_389185       step_1  2017-04-11 08:42:36  
321308  586047816_14599436811_389185        start  2017-04-11 08:42:20  

[177847 rows x 6 columns]

In [21]:
# sorting the date and time column for test 
df_final_test['date_time'] = pd.to_datetime(df_final_test['date_time'])
df_final_test['date'] = df_final_test['date_time'].dt.date
df_final_test['time'] = df_final_test['date_time'].dt.time
df_final_test = df_final_test.drop('date_time', axis=1)

display (df_final_test)

C:\Users\Nicole\AppData\Local\Temp\ipykernel_8692\3256818937.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_final_test['date_time'] = pd.to_datetime(df_final_test['date_time'])
C:\Users\Nicole\AppData\Local\Temp\ipykernel_8692\3256818937.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_final_test['date'] = df_final_test['date_time'].dt.date
C:\Users\Nicole\AppData\Local\Temp\ipykernel_8692\3256818937.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFra

client_id Variation             visitor_id  \
0         9988021      Test   580560515_7732621733   
1         9988021      Test   580560515_7732621733   
2         9988021      Test   580560515_7732621733   
3         9988021      Test   580560515_7732621733   
4         9988021      Test   580560515_7732621733   
...           ...       ...                    ...   
321304    6334360      Test  629124187_65258232847   
321305    6334360      Test  629124187_65258232847   
321306    6334360      Test  629124187_65258232847   
321307    6334360      Test  629124187_65258232847   
321308    6334360      Test  629124187_65258232847   

                            visit_id process_step        date      time  
0       781255054_21935453173_531117       step_3  2017-04-17  15:27:07  
1       781255054_21935453173_531117       step_2  2017-04-17  15:26:51  
2       781255054_21935453173_531117       step_3  2017-04-17  15:19:22  
3       781255054_21935453173_531117       step_2  2017-04-17  15:19:13  
4       781255054_21935453173_531117       step_3  2017-04-17  15:18:04  
...                              ...          ...         ...       ...  
321304  586047816_14599436811_389185       step_2  2017-04-11  08:45:38  
321305  586047816_14599436811_389185       step_1  2017-04-11  08:44:43  
321306  586047816_14599436811_389185       step_2  2017-04-11  08:44:30  
321307  586047816_14599436811_389185       step_1  2017-04-11  08:42:36  
321308  586047816_14599436811_389185        start  2017-04-11  08:42:20  

[177847 rows x 7 columns]

In [22]:
# sorting the date and time column for control

df_final_control['date_time'] = pd.to_datetime(df_final_control['date_time'])
df_final_control['date'] = df_final_control['date_time'].dt.date
df_final_control['time'] = df_final_control['date_time'].dt.time
df_final_control = df_final_control.drop('date_time', axis=1)

display (df_final_control)

C:\Users\Nicole\AppData\Local\Temp\ipykernel_8692\3748609319.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_final_control['date_time'] = pd.to_datetime(df_final_control['date_time'])
C:\Users\Nicole\AppData\Local\Temp\ipykernel_8692\3748609319.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_final_control['date'] = df_final_control['date_time'].dt.date
C:\Users\Nicole\AppData\Local\Temp\ipykernel_8692\3748609319.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice fr

client_id Variation             visitor_id  \
20        4033851   Control  762728880_76361333336   
21        4033851   Control  762728880_76361333336   
22        4033851   Control  762728880_76361333336   
23        4033851   Control  762728880_76361333336   
24        4033851   Control  762728880_76361333336   
...           ...       ...                    ...   
321265     393005   Control  238199953_33257685972   
321266    2908510   Control  814969699_90652851448   
321267    2908510   Control  814969699_90652851448   
321268    2908510   Control  814969699_90652851448   
321269    2908510   Control  814969699_90652851448   

                            visit_id process_step        date      time  
20      949661017_22392791362_127391      confirm  2017-04-05  12:29:03  
21      949661017_22392791362_127391      confirm  2017-04-05  12:29:01  
22      949661017_22392791362_127391      confirm  2017-04-05  12:28:52  
23      949661017_22392791362_127391       step_3  2017-04-05  12:26:08  
24      949661017_22392791362_127391       step_2  2017-04-05  12:24:43  
...                              ...          ...         ...       ...  
321265  579367767_60788307935_908431        start  2017-03-29  15:25:20  
321266   562606085_36368381773_92090        start  2017-05-10  22:57:17  
321267   562606085_36368381773_92090       step_2  2017-05-10  22:56:31  
321268   562606085_36368381773_92090       step_1  2017-05-10  22:56:23  
321269   562606085_36368381773_92090        start  2017-05-10  22:56:20  

[143462 rows x 7 columns]

In [23]:
df_completed_test_users = df_final_test[df_final_test['process_step'] == 'confirm']
df_completed_test_users.sort_valuesby=['visitor_id', 'date', 'time']
df_completed_test_users

C:\Users\Nicole\AppData\Local\Temp\ipykernel_8692\1243189354.py:2: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  df_completed_test_users.sort_valuesby=['visitor_id', 'date', 'time']


client_id Variation             visitor_id  \
15        8320017      Test   39393514_33118319366   
35        1982004      Test  618934751_10082078917   
48        3170143      Test  837774974_91015515080   
73        5887227      Test  448202785_26003228044   
99         438567      Test  694624747_14580652583   
...           ...       ...                    ...   
321219    5969816      Test  470587199_31947711941   
321231    8692087      Test   87619541_16268405483   
321232    8692087      Test   87619541_16268405483   
321295    5230357      Test  832474317_11880812098   
321300    6334360      Test  629124187_65258232847   

                            visit_id process_step        date      time  
15      960651974_70596002104_312201      confirm  2017-04-05  13:10:05  
35       132494692_73738638511_70823      confirm  2017-04-17  12:02:42  
48      758001412_25057963943_219559      confirm  2017-04-04  18:41:44  
73      458489157_53226123692_778999      confirm  2017-04-05  14:52:34  
99      787486191_38787201768_120150      confirm  2017-06-08  16:46:39  
...                              ...          ...         ...       ...  
321219  312260787_57270254471_839042      confirm  2017-04-06  10:10:49  
321231  654226272_50322046833_890618      confirm  2017-03-29  17:29:18  
321232  654226272_50322046833_890618      confirm  2017-03-29  17:28:07  
321295   28245330_61880751416_631586      confirm  2017-04-19  11:24:46  
321300  586047816_14599436811_389185      confirm  2017-04-11  08:47:23  

[25722 rows x 7 columns]

In [24]:
#checking for number of duplicates within the confirm subset of the process_step
duplicates = df_completed_test_users.duplicated(subset=['client_id', 'visitor_id', 'visit_id'])
duplicate_count = duplicates.sum()
duplicate_count

3931

In [25]:
# analyzed only the first stamp of the vistor_id 
df_sorted = df_final_test.sort_values(by=['visitor_id', 'date', 'time'])
df_first_time = df_sorted.drop_duplicates(subset=['time'], keep='first')

completed_test = df_first_time[df_first_time['process_step'] == 'confirm']['visitor_id'].nunique()

total_test = df_first_time['visitor_id'].nunique()

completion_rate = (completed_test / total_test) * 100

print(completion_rate)

36.89387917329094


In [26]:
df_first_time['visitor_id'].nunique()

20128

In [27]:
# defining a funtion
def calculate_completion_rate(df):
    df_sorted = df.sort_values(by=['visitor_id', 'date', 'time'])
    df_first_time = df_sorted.drop_duplicates(subset=['time'], keep='first')
    completed_test = df_first_time[df_first_time['process_step'] == 'confirm']['visitor_id'].nunique()
    total_test = df_first_time['visitor_id'].nunique()
    completion_rate = (completed_test / total_test) * 100
    return f"{completion_rate:.2f}%"

In [28]:
#completed_test_users = df_final_test[df_final_test['process_step'] == 'confirm']
#df_final_test.duplicates(subset=['process_step'] == 'confirm', inplace = True)
#df_final_test.groubby
# create a new column that take one confirm per vistor id and that = True

# client id . nunique of the 
# sort the date and time, time in ascending order, drop if duplicated 
# first time stamp the vistor id reached the confirm step

#visits_with_error_test = df_final_test.groupby('visit_id')['error_step'].any()

In [29]:
total_control_users = df_final_control['visitor_id'].nunique()
total_test_users = df_final_test['visitor_id'].nunique()

COMPLETION RATE KPI

In [30]:
# ​​Completion Rate: The proportion of users who reach the final ‘confirm’ step
# A client is said to him completed the process if he reached the confirm step. 
# For the purrpose of analysis and to compare we took the first time stamp that the client reached the confirm step.
# This rules out duplications for the same client_id

# Steps:
# 1) sort the values of the visitor_id, date, time column 
# 2) keep the first time stamp of confirm per visitor_id and drop the rest
# 3) count the number of unique values = completed test count
# 4) count the number of total visitor_id = total test count
# 5) completion rate = completed test / total_test
# 6) return completion_rate upto two decimals and with a percent sign %


def calculate_completion_rate(df):
    df_sorted = df.sort_values(by=['visitor_id', 'date', 'time'])
    df_first_time = df_sorted.drop_duplicates(subset=['time'], keep='first')
    completed_test = df_first_time[df_first_time['process_step'] == 'confirm']['visitor_id'].nunique()
    total_test = df_first_time['visitor_id'].nunique()
    completion_rate = (completed_test / total_test) * 100
    return f"{completion_rate:.2f}%"


In [31]:
completion_rate_test = calculate_completion_rate(df_final_test)
completion_rate_control = calculate_completion_rate(df_final_control)

print (completion_rate_test)
print (completion_rate_control)

36.89%
34.06%


In [32]:
#Testing the Hypothesis
#H0 completion_rate_test = complettion_rate_contol
#H1 completion_rate_test > completion_rate_control
#alpha = 0.05

import statsmodels.api as sm
from math import sqrt

#test for proportions?? 
n_test = df_final_test['visitor_id'].nunique()
n_control = df_final_control['visitor_id'].nunique()

completion_rate_test = calculate_completion_rate(df_final_test)
completion_rate_control = calculate_completion_rate(df_final_control)

#calculate the number of completions
x_test = completion_rate_test * total_test
x_control = completion_rate_control * total_control

count = np.array([x_test, x_control])
nobs = np.array([n_test, n_control])

z_stat, p_value = sm.stats.proportions_ztest(count, nobs, alternative='larger')



NameError: name 'total_control' is not defined

In [ ]:
# Average time spent on each step

def average_time_spent_on_each_step(df):
    df['date'] = df['date'].astype(str)
    df['time'] = df['time'].astype(str)
    df['datetime'] = pd.to_datetime(df['date'] + ' ' + df['time'])
    df = df.sort_values(by=['visit_id', 'datetime'])
    df['duration'] = df.groupby(['visit_id'])['datetime'].diff(-1).dt.total_seconds()
    df['duration'] = df['duration'].abs()
    average_step_durations = df.groupby('process_step')['duration'].mean()
    average_step_durations_minutes = average_step_durations / 60
    return average_step_durations_minutes

In [ ]:
average_time_spent_on_each_step (df_final_control)

In [ ]:
average_time_spent_on_each_step (df_final_control)

In [ ]:
df_final_control

In [ ]:
average_durations_test_group = average_time_spent_on_each_step(df_final_test)
average_durations_control_group = average_time_spent_on_each_step(df_final_control)

print(average_durations_test_group )
print(average_durations_control_group )

In [ ]:
def plot_average_durations_comparison(test_group, control_group):
    process_steps = test_group.index
    test_durations = test_group.values
    control_durations = control_group.values

    # Set the positions and width for the bars
    bar_width = 0.35
    index = np.arange(len(process_steps))

    # Create the plot
    plt.figure(figsize=(10, 6))

    # Plot both control and test group bars
    plt.bar(index, control_durations, bar_width, label='Control Group', color='lightcoral')
    plt.bar(index + bar_width, test_durations, bar_width, label='Test Group', color='skyblue')

    # Add labels, title, and grid
    plt.xlabel('Process Step')
    plt.ylabel('Average Duration (minutes)')
    plt.title('Comparison of Average Duration by Process Step: Control vs. Test Groups')
    plt.xticks(index + bar_width / 2, process_steps)
    plt.legend()

    # Add gridlines for better readability
    plt.grid(axis='y', linestyle='--', alpha=0.7)

    # Show the plot
    plt.tight_layout()
    plt.show()

# Example usage:
plot_average_durations_comparison(average_durations_test_group, average_durations_control_group)

In [ ]:
#Hypothesis Testing
#GOAL: to assess the efficiency of a process and determine if changes reduce time spent per step.
#H0: average_durations_test_group >= average_durations_control_group
#H1: average_durations_test_group < average_durations_control_group 
#(suggesting that the test group takes significantly less time to complete each step compared to the control group)
    #one-sided so one-tailed t-test 
#alpha = 0.05

#two independent groups, two sample t-test, one-tailed t-test
import scipy.stats as st

average_durations_test_group = [3.971134, 1.009534, 1.009814, 1.476842, 2.148741]  
average_durations_control_group = [2.627372, 1.048043, 0.837757, 1.526548, 2.257719]  

t_stat,p_value = st.ttest_ind(average_durations_test_group,average_durations_control_group, equal_var=False, alternative="less")

alpha = 0.05

print(f"T-statistic: {t_stat}")
print(f"P-value: {p_value}")

if p_value < alpha:
    print("Reject the null hypothesis (H0). There is sufficient evidence to suggest that the test group spends less time on average.")
else:
    print("Accept the null hypothesis (H0). There is insufficient evidence to suggest that the test group spends less time on average.")

Conclusion:
There is insufficient evidence to support the claim that the test group spends significantly less time on each step compared to the control group. In fact, the high p-value suggests that any observed difference in the time spent on each step could easily be due to random chance. Therefore, we cannot conclude that the test group is faster than the control group based on the given data.

In [ ]:
#Hypothesis Testing
#GOAL: to assess the efficiency of a process and determine if changes reduce time spent per step.
#H0: average_durations_test_group <= average_durations_control_group
#H1: average_durations_test_group > average_durations_control_group 
#(suggesting that the test group takes significantly less time to complete each step compared to the control group)
    #one-sided so one-tailed t-test 
#alpha = 0.05

#two independent groups, two sample t-test, one-tailed t-test
import scipy.stats as st

average_durations_test_group = [3.971134, 1.009534, 1.009814, 1.476842, 2.148741]  
average_durations_control_group = [2.627372, 1.048043, 0.837757, 1.526548, 2.257719]  

t_stat,p_value = st.ttest_ind(average_durations_test_group, average_durations_control_group,equal_var=False, alternative="greater")

alpha = 0.05

print(f"T-statistic: {t_stat}")
print(f"P-value: {p_value}")

if p_value < alpha:
    print("Reject the null hypothesis (H0). There is sufficient evidence to suggest that the test group spends less time on average.")
else:
    print("Fail to reject the null hypothesis (H0). There is insufficient evidence to suggest that the test group spends less time on average.")

In [ ]:
df_final_test

In [ ]:
df_final_test['process_step'].unique()

In [ ]:
df_final_test

In [ ]:
# Add a 'group' column to each DataFrame
df_final_test['group'] = 'test'
df_final_control['group'] = 'control'

# Concatenate the DataFrames
df_combined = pd.concat([df_final_test, df_final_control])

# Create the box plot
plt.figure(figsize=(10, 6))
sns.boxplot(x='process_step', y='duration', hue='group', data=df_combined)
plt.title('Box Plot of Durations by Process Step and Group')
plt.xlabel('Process Step')
plt.ylabel('Duration')
plt.xticks(rotation=45)
plt.legend(title='Group')
plt.show()

In [ ]:
import pandas as pd
import numpy as np

# Function to calculate outliers and filter them out
def remove_outliers(df):
    # Calculate Q1 (25th percentile) and Q3 (75th percentile) for each process step
    Q1 = df.groupby('process_step')['duration'].quantile(0.25)
    Q3 = df.groupby('process_step')['duration'].quantile(0.75)
    IQR = Q3 - Q1
    
    # Define bounds for outliers
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    
    # Create a DataFrame for outlier bounds
    bounds = pd.DataFrame({
        'lower_bound': lower_bound,
        'upper_bound': upper_bound
    })
    
    # Merge bounds with the original DataFrame
    df_with_bounds = df.merge(bounds, on='process_step')
    
    # Filter out the outliers
    df_filtered = df_with_bounds[(df_with_bounds['duration'] >= df_with_bounds['lower_bound']) &
                                 (df_with_bounds['duration'] <= df_with_bounds['upper_bound'])]
    
    return df_filtered

# Add a 'group' column to each DataFrame
df_final_test['group'] = 'test'
df_final_control['group'] = 'control'

# Combine the DataFrames
df_combined = pd.concat([df_final_test, df_final_control])

# Remove outliers
df_filtered = remove_outliers(df_combined)

# Create the box plot without outliers
plt.figure(figsize=(12, 8))
sns.boxplot(x='process_step', y='duration', hue='group', data=df_filtered)
plt.title('Box Plot of Durations by Process Step and Group (Outliers Removed)')
plt.xlabel('Process Step')
plt.ylabel('Duration')
plt.xticks(rotation=45)
plt.legend(title='Group')
plt.show()


In [ ]:
# Function to calculate outliers and filter them out
def remove_outliers(df):
    Q1 = df.groupby('process_step')['duration'].quantile(0.25)
    Q3 = df.groupby('process_step')['duration'].quantile(0.75)
    IQR = Q3 - Q1
    
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    
   
    bounds = pd.DataFrame({
        'lower_bound': lower_bound,
        'upper_bound': upper_bound
    })
    
    df_with_bounds = df.merge(bounds, on='process_step')
    
    df_filtered = df_with_bounds[(df_with_bounds['duration'] >= df_with_bounds['lower_bound']) &
                                 (df_with_bounds['duration'] <= df_with_bounds['upper_bound'])]
    
    return df_filtered



In [ ]:
# Function to calculate outliers, remove them, and convert duration to minutes
def remove_outliers_and_convert(df):
    df['duration'] = df['duration'].abs() / 60
    Q1 = df.groupby('process_step')['duration'].quantile(0.25)
    Q3 = df.groupby('process_step')['duration'].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    bounds = pd.DataFrame({'lower_bound': lower_bound,'upper_bound': upper_bound})
    df_with_bounds = df.merge(bounds, on='process_step')
    df_filtered = df_with_bounds[(df_with_bounds['duration'] >= df_with_bounds['lower_bound']) &
                                 (df_with_bounds['duration'] <= df_with_bounds['upper_bound'])]
    
    return df_filtered


In [ ]:
df_final_test = df_final_test.sort_values(by=['visit_id', 'datetime'])
df_final_test['duration'] = df_final_test.groupby(['visit_id'])['datetime'].diff(-1).dt.total_seconds()

df_final_test['duration'] = df_final_test['duration'].abs()

#df_final_test['duration'] = df_final_test['duration'].fillna(0)

df_final_test


In [ ]:
df_final_control = df_final_control.sort_values(by=['visit_id', 'datetime'])
df_final_control['duration'] = df_final_control.groupby(['visit_id'])['datetime'].diff(-1).dt.total_seconds()

df_final_control['duration'] = df_final_control['duration'].abs()

#df_final_test['duration'] = df_final_test['duration'].fillna(0)

df_final_control


In [ ]:
remove_outliers_and_convert (df_final_test)

In [ ]:
remove_outliers_and_convert(df_final_control)

In [ ]:
# Function to calculate outliers, remove them, and convert duration to minutes
def remove_outliers_and_convert(df):
    df['duration'] = df['duration'].abs() / 60
    Q1 = df.groupby('process_step')['duration'].quantile(0.25)
    Q3 = df.groupby('process_step')['duration'].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    bounds = pd.DataFrame({'lower_bound': lower_bound,'upper_bound': upper_bound})
    df_with_bounds = df.merge(bounds, on='process_step')
    df_filtered = df_with_bounds[(df_with_bounds['duration'] >= df_with_bounds['lower_bound']) &
                                 (df_with_bounds['duration'] <= df_with_bounds['upper_bound'])]
    
    return df_filtered

custom_palette = {'Control': '#96151d', 'Test': '#FFCDD2'}
                  

df_final_control['group'] = 'Control'
df_final_test['group'] = 'Test'

df_combined = pd.concat([df_final_test, df_final_control])

# Remove outliers and convert duration to minutes
df_filtered = remove_outliers_and_convert(df_combined)

# Create the box plot without outliers using the custom palette
plt.figure(figsize=(12, 8))
sns.boxplot(x='process_step', y='duration', hue='group', data=df_filtered, palette=custom_palette)
plt.title('Average Time Spent on Each Step by Experiment Groups (Outliers Removed)')
plt.xlabel('Process Step')
plt.ylabel('Duration (minutes)')
plt.xticks(rotation=45)
plt.legend(title='Variation')
plt.savefig("Average Time Spent on Each Step by Experiment Groups.png", transparent=True, bbox_inches='tight')
plt.show()


In [ ]:
# IN SECONDS

def remove_outliers_and_convert(df):
    df['duration'] = df['duration'].abs()
    Q1 = df.groupby('process_step')['duration'].quantile(0.25)
    Q3 = df.groupby('process_step')['duration'].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    bounds = pd.DataFrame({'lower_bound': lower_bound,'upper_bound': upper_bound})
    df_with_bounds = df.merge(bounds, on='process_step')
    df_filtered = df_with_bounds[(df_with_bounds['duration'] >= df_with_bounds['lower_bound']) &
                                 (df_with_bounds['duration'] <= df_with_bounds['upper_bound'])]
    
    return df_filtered

custom_palette = {'test':  '#96151d' , 'control': '#FFCDD2'}
                  
df_final_test['group'] = 'test'
df_final_control['group'] = 'control'

# Combine the DataFrames
df_combined = pd.concat([df_final_test, df_final_control])

# Remove outliers and convert duration to minutes
df_filtered = remove_outliers_and_convert(df_combined)

# Create the box plot without outliers using the custom palette
plt.figure(figsize=(12, 8))
sns.boxplot(x='process_step', y='duration', hue='group', data=df_filtered, palette=custom_palette)
plt.title('Box Plot of Durations by Process Step and Group (Outliers Removed)')
plt.xlabel('Process Step')
plt.ylabel('Duration')
plt.xticks(rotation=45)
plt.legend(title='Group')
plt.show()


In [ ]:
df_filtered


In [ ]:
df_filtered_test = remove_outliers(df_final_test)
df_filtered_control = remove_outliers(df_final_control)

In [ ]:
average_durations_test_group = average_time_spent_on_each_step(df_filtered_test)
print (average_durations_test_group)

In [ ]:
average_durations_test_group = average_time_spent_on_each_step(df_filtered_control)
print (average_durations_test_group)

HYPOTHESIS TESTING - AVERAGE COMPLETION TIME

In [ ]:
#Hypothesis Testing
#GOAL: to assess the efficiency of a process and determine if changes reduce time spent per step.
#H0: average_durations_test_group >= average_durations_control_group
#H1: average_durations_test_group < average_durations_control_group 
#(suggesting that the test group takes significantly less time to complete each step compared to the control group)
    #one-sided so one-tailed t-test 
#alpha = 0.05

#two independent groups, two sample t-test, one-tailed t-test
import scipy.stats as st

average_durations_test_group = [2.054426, 0.881014, 1.054969, 1.604868, 1.663180]  
average_durations_control_group = [1.262612, 1.032584, 0.852407, 1.650102, 1.520628]  

t_stat,p_value = st.ttest_ind(average_durations_test_group,average_durations_control_group, equal_var=False, alternative="less")

alpha = 0.05

print(f"T-statistic: {t_stat}")
print(f"P-value: {p_value}")

if p_value < alpha:
    print("Reject the null hypothesis: average_durations_test_group < average_durations_control_group ")
else:
    print("Accept the null hypothesis: average_durations_test_group >= average_durations_control_group or No significant difference in time between the groups.")


------- EXTRA --------

In [ ]:
#Convert the date and time columns to a single datetime column
#This is so that we have correct sequence and i can accurately gather the time for each day. Also I am just dealing with one column

#Calculate the time spent between steps.
#Aggregate the time spent for each step.
#Compute the mean duration for each step.


In [ ]:
df_final_test['datetime'] = pd.to_datetime(df_final_test['date'] + ' ' + df_final_test['time'])
    


In [ ]:
print (df_final_test.dtypes)

In [ ]:
#change the dataype to string so that we can concat them into one column
df_final_test['date'] = df_final_test['date'].astype(str)
df_final_test['time'] = df_final_test['time'].astype(str)

df_final_test['datetime'] = pd.to_datetime(df_final_test['date'] + ' ' + df_final_test['time'])
    
df_final_test

In [ ]:
#df_final_test = df_final_test.sort_values(by=['visit_id', 'datetime'],ascending=False)
#df_final_test['duration'] = df_final_test.groupby(['visit_id'])['datetime'].diff(-1).dt.total_seconds()
#df_final_test['duration'] = df_final_test['duration'].fillna(0)

#df_final_test


In [ ]:
visit_id_to_check = df_final_test[df_final_test['visit_id'] == '999988789_76411676596_272843']
visit_id_to_check

In [ ]:
average_step_durations = df_final_test.groupby('process_step')['duration'].mean()
average_step_durations # in seconds

In [ ]:
average_step_durations_sorted = average_step_durations.sort_values(ascending=False)
average_step_durations_sorted

In [ ]:
average_step_durations_minutes = average_step_durations / 60
average_step_durations_minutes

In [ ]:
process_steps = ['confirm', 'start', 'step_1', 'step_2', 'step_3']
average_durations_minutes = [3.971134, 1.009534, 1.009814, 1.476842, 2.148741]

plt.figure(figsize=(10, 6))
plt.barh(process_steps, average_durations_minutes, color='skyblue')
plt.xlabel('Average Duration (minutes)')
plt.ylabel('Process Step')
plt.title('Average Duration of Each Process Step')
plt.gca().invert_yaxis()  # Optional: invert y-axis to have the highest value on top
plt.grid(axis='x', linestyle='--', alpha=0.7)

plt.show()

In [ ]:
visit_id_to_check = df_final_test[df_final_test['visit_id'] == '999988789_76411676596_272843']
visit_id_to_check

In [ ]:
from statsmodels.stats.proportion import proportions_ztest

#H0 = error rate test group >= error rate control group
#H1 = error rate test group < error rate control group

def calculate_errors_and_visits(df):
    df['process_step_number'] = df['process_step'].replace({'start': 0, 'step_1': 1,'step_2': 2,'step_3': 3,'confirm': 4})
    df['step_diff'] = df.groupby('visit_id')['process_step_number'].diff(-1)
    count_errors = df[df['step_diff'] < 0]['step_diff'].count()
    total_visits = df['visit_id'].nunique()  # Number of unique visits
    return count_errors, total_visits

# Calculate errors and visits for both groups
errors_test, visits_test = calculate_errors_and_visits(df_final_test)
errors_control, visits_control = calculate_errors_and_visits(df_final_control)

# Z-test for proportions
count = [errors_test, errors_control]
nobs = [visits_test, visits_control]
stat, p_value = proportions_ztest(count, nobs, alternative='smaller')  # One-tailed test

print(f"Z-statistic: {stat}, P-value: {p_value}")

# print the test result
if p_value < 0.05:
    print('The null hypothesis is rejected. The error rate in the test group is lower than the control group.')
else:
    print('We do not have sufficient proof to reject null hypothesis. We cannot confirm the error rate in the test group is lower than in the control group.')

print('P-value of one might suggest that the error rates between the groups are either identical or very close, leading to a very small difference being inflated by the large sample sizes.')

------------ EXTRA ----------

------ COMPLETION RATE KPI --------

In [ ]:
# ​​Completion Rate: The proportion of users who reach the final ‘confirm’ step
# A client is said to him completed the process if he reached the confirm step. 
# For the purrpose of analysis and to compare we took the first time stamp that the client reached the confirm step.
# This rules out duplications for the same client_id

# Steps:
# 1) sort the values of the visitor_id, date, time column 
# 2) keep the first time stamp of confirm per visitor_id and drop the rest
# 3) count the number of unique values = completed test count
# 4) count the number of total visitor_id = total test count
# 5) completion rate = completed test / total_test
# 6) return completion_rate upto two decimals and with a percent sign %


def calculate_completion_rate(df):
    df_sorted = df.sort_values(by=['visitor_id', 'date', 'time'])
    df_first_time = df_sorted.drop_duplicates(subset=['time'], keep='first')
    completed_test = df_first_time[df_first_time['process_step'] == 'confirm']['visitor_id'].nunique()
    total_test = df_first_time['visitor_id'].nunique()
    completion_rate = (completed_test / total_test) * 100
    return f"{completion_rate:.2f}%"




In [ ]:
completion_rate_test = calculate_completion_rate(df_final_test)
completion_rate_control = calculate_completion_rate(df_final_control)

print (completion_rate_test)
print (completion_rate_control)

In [ ]:
from statsmodels.stats.proportion import proportions_ztest

#H0 = completion rate test group >= completion rate control group
#H1 = completion rate test group < completion rate control group

def calculate_completion_counts(df):
    df_sorted = df.sort_values(by=['visitor_id', 'date', 'time'])
    df_first_time = df_sorted.drop_duplicates(subset=['time'], keep='first')
    completed = df_first_time[df_first_time['process_step'] == 'confirm']['visitor_id'].nunique()
    total = df_first_time['visitor_id'].nunique()
    return completed, total


completed_test, total_test = calculate_completion_counts(df_final_test)
completed_control, total_control = calculate_completion_counts(df_final_control)


count = [completed_test, completed_control]
nobs = [total_test, total_control]
stat, p_value = proportions_ztest(count, nobs, alternative='smaller')  # One-tailed test

print(f"Z-statistic: {stat}, P-value: {p_value}")

# Print the test result based on the p-value
if p_value < 0.05:
    print('The null hypothesis is rejected. The completion rate in the test group is significantly lower than in the control group.')
else:
    print('Accept the null hypothesis. We cannot confirm that the completion rate in the test group is lower than in the control group.')


completion_rate_test = (completed_test / total_test) * 100
completion_rate_control = (completed_control / total_control) * 100
print(f"Completion rate (Test group): {completion_rate_test:.2f}%")
print(f"Completion rate (Control group): {completion_rate_control:.2f}%")


Interpretation: Accept the null hypothesis. The test group’s completion rate is likely higher than or at least equivalent to that of the control group, based on these results.

Part2: Assessing the Practical Significance with a 5% Threshold

In [ ]:
#H0 = The increase in the Test group’s completion rate is less than 5%
#H1 = The increase in the Test group’s completion rate is greater than or equal to 5%.

from statsmodels.stats.proportion import proportions_ztest

# Calculate completion counts and total visits for both groups
completed_test, total_test = calculate_completion_counts(df_final_test)
completed_control, total_control = calculate_completion_counts(df_final_control)

# Observed completion rates
completion_rate_test = completed_test / total_test
completion_rate_control = completed_control / total_control

# Z-test for difference in completion rates (Test group vs Control group)
count = [completed_test, completed_control]
nobs = [total_test, total_control]
stat, p_value = proportions_ztest(count, nobs, alternative='larger')  # One-tailed test

print(f"Z-statistic: {stat}, P-value: {p_value}")
print(f"Completion rate (Test group): {completion_rate_test:.4f}")
print(f"Completion rate (Control group): {completion_rate_control:.4f}")

# Print the result of the hypothesis test
if p_value < 0.05:
    print('The null hypothesis is rejected. The completion rate in the Test group is significantly higher than in the Control group.')
else:
    print('We do not have sufficient evidence to reject the null hypothesis.')

# Part 2: Assess if the increase meets or exceeds the 5% threshold

# Minimum increase to justify the cost
threshold_increase = 0.05

# Observed difference in completion rates
observed_difference = completion_rate_test - completion_rate_control

# Check if the observed increase meets or exceeds the 5% threshold
if observed_difference >= threshold_increase:
    print(f"The observed increase of {observed_difference:.4f} meets or exceeds the 5% threshold. The new design is cost-effective.")
else:
    print(f"The observed increase of {observed_difference:.4f} does not meet the 5% threshold. The new design may not be cost-effective.")


Interpretation of Results:
1. Statistical Significance:
Z-statistic: 5.86 and P-value: 2.29e-09:
The Z-statistic is very high, and the P-value is extremely small, much lower than the typical 0.05 significance level.
Conclusion: We reject the null hypothesis. The completion rate in the Test group (new design) is significantly higher than in the Control group (old design).
Completion Rate (Test group): 36.89% vs. Completion Rate (Control group): 34.06%:
The Test group has a higher completion rate, and the difference of 2.83% is statistically significant.

2. Practical Significance (5% Threshold):
Observed Increase: 2.83%:
Vanguard's threshold for cost-effectiveness is a minimum 5% increase in completion rate. The observed increase of 2.83% is below the 5% threshold.
Conclusion: Despite the statistically significant improvement in completion rate, the new design does not meet the 5% threshold required to justify the associated costs of the redesign (development, design, and potential user disruption).
Business Implications:
Statistically Significant, But Not Cost-Effective:

The new design performs better than the old one, but not by a margin that justifies the cost of implementation. 
The observed increase of 2.83% is below the 5% minimum set by Vanguard for a cost-effective upgrade.


Decision-Making:

From a statistical perspective: The new design improves user completion rates, which is a positive indicator for user engagement.
From a business perspective: Given that the improvement does not reach the 5% threshold, Vanguard may decide not to proceed with the full implementation of the new design unless additional benefits (beyond completion rate) can be identified to offset the costs.

Next Steps:

Further Optimization: Consider additional design improvements or A/B testing to explore whether further changes can push the increase beyond the 5% threshold.
Cost-Benefit Reassessment: Reassess the cost implications and potential non-monetary benefits (e.g., user satisfaction, long-term engagement) to determine if the design is worth rolling out despite not meeting the completion rate threshold.
In summary, while the new design is statistically effective, it may not yet be financially justifiable. Vanguard should weigh the potential benefits against the implementation costs before making a final decision.
ive.

Vanguard wants the users to reach the confirm page ---- so completion rate is the golden KPI